### PR TITLE
Upgrade Jest from 27 to 29.5.0

### PR DESCRIPTION
**Pull request by jsnajdr**
This is an upgrade of Jest from v27 to v29, two major versions up. There's been a lot of breaking changes, so it's a big and difficult PR.

At this moment, web unit tests are all green, mobile tests still need some fixing.

One thing that's broken in mobile tests are Jest helpers imported from `react-native-reanimated`. Gutenberg uses a [fork of `react-native-reanimated`](https://github.com/wordpress-mobile/react-native-reanimated). For the Jest update, we need to incorporate changes from this PR: https://github.com/software-mansion/react-native-reanimated/pull/3559. @geriux would it be possible to update the fork? Currently the fork is based on upstream version 2.9.1 (Jul 2022), the Jest helper issue (`expect.extend`) has been fixed in version [2.13.0](https://github.com/software-mansion/react-native-reanimated/releases/tag/2.13.0) (Nov 2022) and the latest version is 2.14.4 (Jan 2022, two weeks ago)

### CODE REVIEW

**Comment by geriux** [comment]
> @geriux would it be possible to update the fork? Currently the fork is based on upstream version 2.9.1 (Jul 2022), the Jest helper issue (expect.extend) has been fixed in version [2.13.0](https://github.com/software-mansion/react-native-reanimated/releases/tag/2.13.0) (Nov 2022) and the latest version is 2.14.4 (Jan 2022, two weeks ago)

I think what we could do is cherry-pick that [change](https://github.com/software-mansion/react-native-reanimated/pull/3559/files) into the fork, that would be quicker than updating to a newer version as that requires more work and a lot of testing.

**Comment by jsnajdr** [comment]
> I think what we could do is cherry-pick that change into the fork

That would be awesome. The issue is that Jest's `expect` package used to be CommonJS, and `require('extend')` used to return the actual `extend` object, but now it's an ESM module and you need to use `require('extend').default` instead.

**Comment by geriux** [comment]
> That would be awesome. The issue is that Jest's `expect` package used to be CommonJS, and `require('extend')` used to return the actual `extend` object, but now it's an ESM module and you need to use `require('extend').default` instead.

Cool! I'll find some time to cherry-pick that and bump the Reanimated version, I'll let you know once it is ready 👍

**Comment by geriux** [comment]
Hey @jsnajdr 👋  I'm working on a [draft PR](https://github.com/WordPress/gutenberg/pull/47574) that will update the Reanimated version. In the meantime, you could try pointing to [this version](https://github.com/WordPress/gutenberg/pull/47574/files#diff-1f256578dd9ab8dda6c3913a55a588312129f3404bb0cdf0b1ab52f0a6937119R69) and see if this issue is resolved. Thanks!

**Comment by geriux** [comment]
Hey there 👋 The Reanimated update is now merged in `trunk`.

**Comment by jsnajdr** [comment]
> The Reanimated update is now merged in trunk.

Thanks for the update @geriux, it really fixed a lot of failures in this branch.

Turns out we need to cherry-pick one more patch for `react-native-reanimated`. There is the `react-native-reanimated/lib/index.js` file that contains an "export namespace as default" statement:
```js
export * as default from './Animated';
```
This syntax is however not supported without an extra Babel plugin, so we need to rewrite it to:
```js
import * as Animated from './Animated';
export default Animated;
```
In the upstream `react-native-reanimated` library this was done in https://github.com/software-mansion/react-native-reanimated/pull/3437 (look just at the `src/index.ts` change), and the change is included in recent versions like 2.14.4.

**Comment by jsnajdr** [comment]
All CI checks are green now :+1: Ready for review and merging.

**Comment by geriux** [comment]
> Turns out we need to cherry-pick one more patch for `react-native-reanimated`. There is the `react-native-reanimated/lib/index.js` file that contains an "export namespace as default" statement:

Hey there @jsnajdr 👋 sorry for the delay, is this still needed? I see all checks are passing 🟢. Let us know so we can help. Thank you!

**Comment by jsnajdr** [comment]
> sorry for the delay, is this still needed? I see all checks are passing

I would be still nice to have, as newer upstream versions have that patch. But it's no longer necessary. For this branch I solved it by adding an explicit Babel config file for `packages/e2e-tests`: https://github.com/WordPress/gutenberg/pull/47388/commits/535303795e317fb84d3db6601b675f2cea8dafbd

The `babel-preset-default` adds the `export-default-from` transform, needed to support this syntax, and also the `reanimated` Babel plugin which transforms the Reanimated worklets and is needed for animations to work.

**Comment by geriux** [comment]
> I would be still nice to have, as newer upstream versions have that patch.

I see, we'll plan to bring that then into a patch release of our fork 👍 

>For this branch I solved it by adding an explicit Babel config file for packages/e2e-tests: https://github.com/WordPress/gutenberg/commit/535303795e317fb84d3db6601b675f2cea8dafbd

Awesome, glad to know that's not blocking this PR.

File: packages/compose/src/higher-order/with-global-events/test/index.js
Comment (tyxla) [inline packages/compose/src/higher-order/with-global-events/test/index.js]: I'm curious if this change makes any difference.

File: packages/jest-preset-default/jest-preset.js
Comment (tyxla) [inline packages/jest-preset-default/jest-preset.js:27]: Nice catch 👍 Seems like we missed that when we defaulted to real timers.

File: packages/scripts/scripts/test-e2e.js
Comment (tyxla) [inline packages/scripts/scripts/test-e2e.js]: The `--runInBand` change seems to deserve its own PR, doesn't it? Or is it for some reason strictly related to this one?

File: packages/jest-preset-default/package.json
Comment (tyxla) [inline packages/jest-preset-default/package.json]: This is worth a CHANGELOG entry.

File: packages/jest-console/package.json
Comment (tyxla) [inline packages/jest-console/package.json]: This is worth a CHANGELOG entry.

File: packages/react-native-editor/src/test/block-parser-code.test.js
Comment (tyxla) [inline packages/react-native-editor/src/test/block-parser-code.test.js]: I suppose those imports are unnecessary because of:

https://github.com/WordPress/gutenberg/blob/705ab44f2c2797d75740533a85030c194b4fb040/packages/react-native-editor/src/index.js#L10

File: packages/report-flaky-tests/package.json
Comment (tyxla) [inline packages/report-flaky-tests/package.json]: This package's CHANGELOG is quite empty, but we still could include an entry for this change there.

File: packages/scripts/package.json
Comment (tyxla) [inline packages/scripts/package.json]: Those changes are worth one (or even multiple) CHANGELOG entries. This package is used broadly externally, so we should be as thorough as possible when introducing major changes like this.

File: packages/e2e-tests/package.json
Comment (tyxla) [inline packages/e2e-tests/package.json]: We could add a CHANGELOG entry here.

File: packages/e2e-tests/babel.config.js
Comment (tyxla) [inline packages/e2e-tests/babel.config.js:2]: Nice 👍 

I can't think of a scenario where we'd actually want to cache the config per environment with `api.cache.using( () => process.env.NODE_ENV )`.

I wonder if enabling this config cache was motivated by something and whether this change affects the performance of the e2e tests.

File: package.json
Comment (tyxla) [inline package.json]: Should we also bump some peer deps, like here:

https://github.com/WordPress/gutenberg/commit/caab0febf95760b6e4de61cc7663d0d7fa5c149d

I've noticed that being very flexible with dependencies can cause issues with more complex upgrades (upgrading `@wordpress` packages for example). Maybe it won't be the case with `jest` though.

File: packages/scripts/utils/test/index.js
Comment (tyxla) [inline packages/scripts/utils/test/index.js]: Why was this IIFE necessary?

File: test/native/jest.config.js
Comment (tyxla) [inline test/native/jest.config.js:28]: Nice simplification here 👍 

It seems like it's arguable which one is better since now we have to specify a bunch of additional traversing below. `../../`.

File: test/native/jest.config.js
Comment (tyxla) [inline test/native/jest.config.js]: Do we still need to keep these ignore patterns? Or is there a good reason we're removing them?

File: test/unit/scripts/resolver.js
Comment (tyxla) [inline test/unit/scripts/resolver.js]: Thanks for the thorough explanation 👍

**Review by tyxla — COMMENTED** [review]
Thanks for the great work on this one @jsnajdr 🙌 

I think we're missing quite a few CHANGELOG entries. This is a big upgrade, breaking for some of the packages.

Also, I didn't see an obvious difference when running the tests locally, I wonder if you observed a performance improvement by using a newer `jest` with a newer `jsdom`.

File: packages/compose/src/higher-order/with-global-events/test/index.js
Comment (jsnajdr) [reply packages/compose/src/higher-order/with-global-events/test/index.js]: Without this trick, Jest would hoist the `jest.spyOn` calls above the `super()` call, causing an error. For the same reasons I had to wrap `jest.spyOn` calls in IIFE in other files.

It is a Jest bug, which I reported in https://github.com/facebook/jest/pull/13952 and which I'll try to fix. Then we can revert this part of the patch again.

File: test/native/jest.config.js
Comment (jsnajdr) [reply test/native/jest.config.js:28]: I had to change the root, and also to introduce a special `babel.config.js` in `packages/e2e-tests`, because of https://github.com/facebook/jest/pull/12689. After that change Jest (namely `babel-jest`) will find `test/native/babel.config.js` only if the `rootDir` is `test/native`. So I had to change it, although it introduces extra `../..` traversals.

File: test/native/jest.config.js
Comment (jsnajdr) [reply test/native/jest.config.js]: I'm removing them because:
1. They don't work with `../..` added, like in `<rootDir>/../../.*/build-module`. On regular expression patterns, Jest won't do the path resolution that would resolve `a/b/c/../..` into `a`, and the patterns with `../..` not removed won't match anything.
2. These ignore patterns were a workaround for the fact that there are unit tests in `packages/react-native-editor/src/parser` directory, and that directory is not excluded from build. I solved that by moving them into `packages/react-native-editor/src/test`. After that, these ignores are no longer needed.

File: packages/scripts/scripts/test-e2e.js
Comment (gziolo) [reply packages/scripts/scripts/test-e2e.js]: Does it mean you can safely run e2e tests in parallel now? I'm curious about what has changed in Jest about that. The challenge we had at the beginning was that some tests could update the shared database in a way that gives false negatives when running assertions.

File: packages/react-native-editor/src/test/block-parser-code.test.js
Comment (jsnajdr) [reply packages/react-native-editor/src/test/block-parser-code.test.js]: Yes, the globals are already imported elsewhere.

File: packages/e2e-tests/babel.config.js
Comment (gziolo) [reply packages/e2e-tests/babel.config.js:2]: It's for the test suites' source code. Isn't it always equal to `test`?

Comment (jsnajdr) [reply packages/e2e-tests/babel.config.js:2]: I admin I didn't think about proper caching at all when writing this 🙂 I just copied a statement that every other Babel config in the repo also uses.

Comment (tyxla) [reply packages/e2e-tests/babel.config.js:2]: > It's for the test suites' source code. Isn't it always equal to test?

Yup, that's what I essentially was saying with:

> I can't think of a scenario where we'd actually want to cache the config per environment with `api.cache.using( () => process.env.NODE_ENV )`.

File: test/unit/scripts/resolver.js
Comment (SimenB) [inline test/unit/scripts/resolver.js]: note that uuid v9 does expose CJS entrypoint for browser env.

You might still need this custom resolver, just pointing out that the comment is somewhat stale 🙂

File: packages/scripts/scripts/test-e2e.js
Comment (jsnajdr) [reply packages/scripts/scripts/test-e2e.js]: Oh, it seems that I misunderstood the original code. I thought that
```js
const runInBand = ! hasRunInBand ? [ '--runInBand' ] : [];
```
is a bug, but now I realize it's there to always enforce `--runInBand`, even if it wasn't passed on command line?

I'll fix that, and submit these `test:performance:debug` changes as a standalone PR.

Comment (gziolo) [reply packages/scripts/scripts/test-e2e.js]: Yes, that was the intent but somehow tests pass on CI so I’d like to know how. Maybe it’s passed explicitly there elsewhere but I can’t locate it.

File: test/unit/scripts/resolver.js
Comment (jsnajdr) [reply test/unit/scripts/resolver.js]: Thanks, I see the browser export was fixed in https://github.com/uuidjs/uuid/pull/642. I'll change the comment to mention "uuid v8" explicitly. After we upgrade to uuid v9, we can remove this workaround. I'd rather do the v9 upgrade in a separate PR, I'm not even sure if it will be feasible, as there are also 3rd party libraries which can use old version of `uuid`.

**Comment by jsnajdr** [comment]
I did some final updates for the PR:
- when I started working on this in January, the latest Jest version was 29.3.1, today it's 29.5.0. I updated the PR to upgrade to 29.5.0. One of the benefits is that 29.5.0 fixes a hoisting bug we identified during this upgrade, and I could remove the "avoid unwanted hoisting" workarounds.
- updated peer dependencies of all relevant packages to `jest >= 29`
- added changelog entries to packages where the Jest upgrade is a breaking change

@tyxla @gziolo If the CI checks come back as green, could you do a final review of this upgrade?

File: packages/jest-preset-default/jest-preset.js
Comment (gziolo) [reply packages/jest-preset-default/jest-preset.js:27]: We should also update the README to reflect it:

Screenshot 2023-03-08 at 12 58 44

**Review by gziolo — APPROVED** [review]
Looks good to me. I don't have the knowledge to review some of the changes related to React Native and the custom resolver that seems to be targeting ES Modules in `node_modules`. However, everything is green on CI and the part for the web looks excellent. Great work on that upgrade!

**Comment by jsnajdr** [comment]
> the custom resolver that seems to be targeting ES Modules in `node_modules`.

This works around an issue where the new Jest tries to use the `exports` field in `package.json` and in JSDOM environment, loads the browser version of the package. But some packages export an ESM module as the browser version, instead of CJS, and Jest can't handle that.

The fixes I did were inspired by this very thoughtful comment in another project that did the Jest 29 upgrade last year: https://github.com/microsoft/accessibility-insights-web/pull/5421#issuecomment-1109168149

**Review by tyxla — COMMENTED** [review]
Great work, thank you @jsnajdr 🙌

**Comment by jsnajdr** [comment]
> We should also update the README to reflect it:

@gziolo The `jest-preset-default` README is being updated in #48925. It needed more updates than just the `timers`.

**Comment by jsnajdr** [comment]
After merge, CI checks are failing because `npm ci` is failing for some reason. I'm investigating this.

**Comment by gziolo** [comment]
> After merge, CI checks are failing because `npm ci` is failing for some reason. I'm investigating this.

Most likely nothing to worry about. It often happens after updating the lock file. I believe it's a cache issue and restarting failed jobs helps. See https://github.com/WordPress/gutenberg/issues/33424 for some similar issues in the past.

**Comment by jsnajdr** [comment]
> Most likely nothing to worry about.

Happy to hear that 🙂 Locally, running `npm ci` succeeds, even on clean tree after `distclean`, and neither `npm ci` nor `npm i` produce any additional lockfile changes. There's nothing I could fix.

**Comment by ktmn** [comment]
Updated my `@wordpress/scripts` from `25.2.0` to `26.3.0`, running `wp-scripts test-unit-js` no longer works.

```
Test suite failed to run

    Jest encountered an unexpected token

    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • If you are trying to use TypeScript, see https://jestjs.io/docs/getting-started#using-typescript
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation

    Details:

    /.../my-project/node_modules/.pnpm/uuid@8.3.2/node_modules/uuid/dist/esm-browser/index.js:1
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,jest){export { default as v1 } from './v1.js';
                                                                                      ^^^^^^

    SyntaxError: Unexpected token 'export'

       8 | } from '@wordpress/data'
       9 |
    > 10 | import {
         | ^
      11 |      store as editPostStore,
      12 | } from '@wordpress/edit-post'
      13 |

      at Runtime.createScriptFromCode (../../node_modules/.pnpm/jest-runtime@29.5.0/node_modules/jest-runtime/build/index.js:1495:14)
      at Object.<anonymous> (../../node_modules/.pnpm/@wordpress+blocks@12.9.0_react@18.2.0/node_modules/@wordpress/blocks/build/api/@wordpress/blocks/src/api/factory.js:4:1)
      at Object.<anonymous> (../../node_modules/.pnpm/@wordpress+blocks@12.9.0_react@18.2.0/node_modules/@wordpress/blocks/build/api/@wordpress/blocks/src/api/utils.js:20:1)
      at Object.<anonymous> (../../node_modules/.pnpm/@wordpress+blocks@12.9.0_react@18.2.0/node_modules/@wordpress/blocks/build/store/@wordpress/blocks/src/store/reducer.js:15:1)
      at Object.<anonymous> (../../node_modules/.pnpm/@wordpress+blocks@12.9.0_react@18.2.0/node_modules/@wordpress/blocks/build/store/@wordpress/blocks/src/store/index.js:9:1)
      at Object.<anonymous> (../../node_modules/.pnpm/@wordpress+blocks@12.9.0_react@18.2.0/node_modules/@wordpress/blocks/build/@wordpress/blocks/src/index.js:11:1)
      at Object.<anonymous> (../../node_modules/.pnpm/@wordpress+edit-post@7.9.0_@babel+core@7.21.5_@types+react@18.2.0_react-dom@18.2.0_react@18.2.0/node_modules/@wordpress/edit-post/build/@wordpress/edit-post/src/index.js:4:1)
      ...
```

**Comment by jsnajdr** [comment]
Hi @ktmn :wave: You're running into an issue with `uuid`, discussed in https://github.com/WordPress/gutenberg/pull/47388#discussion_r1118596522 and https://github.com/WordPress/gutenberg/pull/47388#discussion_r1118681987, and described in this code comment:

https://github.com/WordPress/gutenberg/blob/trunk/test/unit/scripts/resolver.js#L10-L22

One way how to solve it is to upgrade `uuid` in your project to v9. But you might not be able to do that, as it's often a transitive dependency of a package you don't have control over.

If that doesn't work, we should probably include the custom Jest resolver that works around these issues in the Jest config for `wp-scripts test-unit-js`.

**Pull request merged**
